src/app/clusters/soil-measurement-server/soil-measurement-server.cpp

```diff
@@ -0,0 +1,124 @@
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+#include <clusters/SoilMeasurement/Metadata.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::SoilMeasurement;
+using namespace chip::app::Clusters::SoilMeasurement::Attributes;
+
+static const Globals::Structs::MeasurementAccuracyRangeStruct::Type kDefaultSoilMoistureMeasurementLimitsAccuracyRange[] = {
+    { .rangeMin = 0, .rangeMax = 100, .percentMax = MakeOptional(static_cast<chip::Percent100ths>(10)) }
+};
+
+static const Globals::Structs::MeasurementAccuracyStruct::Type kDefaultSoilMoistureMeasurementLimits = {
+    .measurementType  = Globals::MeasurementTypeEnum::kSoilMoisture,
+    .measured         = true,
+    .minMeasuredValue = 0,
+    .maxMeasuredValue = 100,
+    .accuracyRanges   = DataModel::List<const Globals::Structs::MeasurementAccuracyRangeStruct::Type>(
+        kDefaultSoilMoistureMeasurementLimitsAccuracyRange)
+};
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace SoilMeasurement {
+
+Instance::Instance(EndpointId aEndpointId) :
+    AttributeAccessInterface(Optional<EndpointId>(aEndpointId), Id), mEndpointId(aEndpointId)
+{}
+
+Instance::~Instance()
+{
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+}
+
+CHIP_ERROR Instance::Init()
+{
+    VerifyOrDie(emberAfContainsServer(mEndpointId, Id) == true);
+
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INCORRECT_STATE);
+
+    // Initialize the soil moisture measurement limits to default values
+    mSoilMeasurementData.soilMoistureMeasurementLimits = kDefaultSoilMoistureMeasurementLimits;
+    mSoilMeasurementData.soilMoistureMeasuredValue.SetNull();
+
+    return CHIP_NO_ERROR;
+}
+
+// This function is intended for the application to set the soil measurement accuracy limits to the proper values during init.
+// Given the limits are fixed, it is not intended to be changes at runtime and the application should call this function only once
+// during init, where changes should not be reported. In case this does have to change during runtime, the application should call
+// this function with reportChange set to true to notify the change. Please not that in such a case, the application should also
+// take care of the ConfigurationVersion attribute to ensure the value is updated.
+CHIP_ERROR
+Instance::SetSoilMeasurementAccuracy(const Globals::Structs::MeasurementAccuracyStruct::Type & measurementLimits, bool reportChange)
+{
+    mSoilMeasurementData.soilMoistureMeasurementLimits = measurementLimits;
+
+    if (reportChange)
+    {
+        MatterReportingAttributeChangeCallback(mEndpointId, SoilMeasurement::Id, SoilMoistureMeasurementLimits::Id);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR
+Instance::SetSoilMeasuredValue(const Attributes::SoilMoistureMeasuredValue::TypeInfo::Type & soilMoistureMeasuredValue)
```

Comment:
I suppose this is more of a spec question, but is this a Q candidate? How much thrashing is expected on this attribute?

Reply:
It is, generally this falls into the measurement category of how often do we expect measurements to be reported by devices. For now we have not marked any of these with Q.